File: API/0338_Adaptive_Bollinger_Breakout/CS/AdaptiveBollingerBreakoutStrategy.cs
using System;
using System.Collections.Generic;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies
{
	/// <summary>
	/// Strategy that trades based on breakouts of Bollinger Bands with adaptively adjusted parameters
	/// based on market volatility.
	/// </summary>
	public class AdaptiveBollingerBreakoutStrategy : Strategy
	{
		private readonly StrategyParam<int> _minBollingerPeriod;
		private readonly StrategyParam<int> _maxBollingerPeriod;
		private readonly StrategyParam<decimal> _minBollingerDeviation;
		private readonly StrategyParam<decimal> _maxBollingerDeviation;
		private readonly StrategyParam<int> _atrPeriod;
		private readonly StrategyParam<DataType> _candleType;

		private int _currentBollingerPeriod;
		private decimal _currentBollingerDeviation;
		private BollingerBands _bollinger;
		private AverageTrueRange _atr;

		private decimal _atrSum;
		private int _atrCount;

		/// <summary>
		/// Strategy parameter: Minimum Bollinger period.
		/// </summary>
		public int MinBollingerPeriod
		{
			get => _minBollingerPeriod.Value;
			set => _minBollingerPeriod.Value = value;
		}

		/// <summary>
		/// Strategy parameter: Maximum Bollinger period.
		/// </summary>
		public int MaxBollingerPeriod
		{
			get => _maxBollingerPeriod.Value;
			set => _maxBollingerPeriod.Value = value;
		}

		/// <summary>
		/// Strategy parameter: Minimum Bollinger deviation.
		/// </summary>
		public decimal MinBollingerDeviation
		{
			get => _minBollingerDeviation.Value;
			set => _minBollingerDeviation.Value = value;
		}

		/// <summary>
		/// Strategy parameter: Maximum Bollinger deviation.
		/// </summary>
		public decimal MaxBollingerDeviation
		{
			get => _maxBollingerDeviation.Value;
			set => _maxBollingerDeviation.Value = value;
		}

		/// <summary>
		/// Strategy parameter: ATR period for volatility calculation.
		/// </summary>
		public int AtrPeriod
		{
			get => _atrPeriod.Value;
			set => _atrPeriod.Value = value;
		}

		/// <summary>
		/// Strategy parameter: Candle type.
		/// </summary>
		public DataType CandleType
		{
			get => _candleType.Value;
			set => _candleType.Value = value;
		}

		/// <summary>
		/// Constructor.
		/// </summary>
		public AdaptiveBollingerBreakoutStrategy()
		{
			_minBollingerPeriod = Param(nameof(MinBollingerPeriod), 10)
			.SetGreaterThanZero()
			.SetDisplay("Min Bollinger Period", "Minimum period for adaptive Bollinger Bands", "Indicator Settings");

			_maxBollingerPeriod = Param(nameof(MaxBollingerPeriod), 30)
			.SetGreaterThanZero()
			.SetDisplay("Max Bollinger Period", "Maximum period for adaptive Bollinger Bands", "Indicator Settings");

			_minBollingerDeviation = Param(nameof(MinBollingerDeviation), 1.5m)
			.SetGreaterThanZero()
			.SetDisplay("Min Bollinger Deviation", "Minimum standard deviation multiplier", "Indicator Settings");

			_maxBollingerDeviation = Param(nameof(MaxBollingerDeviation), 2.5m)
			.SetGreaterThanZero()
			.SetDisplay("Max Bollinger Deviation", "Maximum standard deviation multiplier", "Indicator Settings");

			_atrPeriod = Param(nameof(AtrPeriod), 14)
			.SetGreaterThanZero()
			.SetDisplay("ATR Period", "Period for ATR volatility calculation", "Indicator Settings");

			_candleType = Param(nameof(CandleType), TimeSpan.FromMinutes(5).TimeFrame())
			.SetDisplay("Candle Type", "Type of candles to use", "General");
		}

		/// <inheritdoc />
		public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
		{
			return [(Security, CandleType)];
		}

		/// <inheritdoc />
		protected override void OnReseted()
		{
			base.OnReseted();

			_currentBollingerPeriod = MaxBollingerPeriod; // Start with maximum period
			_currentBollingerDeviation = MinBollingerDeviation; // Start with minimum deviation
			_atr = default;
			_bollinger = default;
			_atrSum = default;
			_atrCount = default;
		}

		protected override void OnStarted(DateTimeOffset time)
		{
			base.OnStarted(time);

			// Create ATR indicator for volatility measurement
			_atr = new AverageTrueRange
			{
				Length = AtrPeriod
			};

			// Create Bollinger Bands indicator with initial parameters
			_bollinger = new BollingerBands
			{
				Length = _currentBollingerPeriod,
				Width = _currentBollingerDeviation
			};

			// Create subscription for candles
			var subscription = SubscribeCandles(CandleType);

			// Bind indicators to subscription and start
			subscription
			.BindEx(_atr, _bollinger, ProcessIndicators)
			.Start();

			// Add chart visualization
			var area = CreateChartArea();
			if (area != null)
			{
				DrawCandles(area, subscription);
				DrawIndicator(area, _bollinger);
				DrawOwnTrades(area);
			}

			// Start position protection with ATR-based stop-loss
			StartProtection(
			takeProfit: new Unit(0), // No fixed take profit
			stopLoss: new Unit(2, UnitTypes.Absolute) // 2 ATR stop-loss
			);
		}

		private void ProcessIndicators(ICandleMessage candle, IIndicatorValue atrValue, IIndicatorValue bollingerValue)
		{
			// Skip unfinished candles
			if (candle.State != CandleStates.Finished)
			return;

			// --- ATR logic (was ProcessAtr) ---
			if (atrValue.IsFinal)
			{
				decimal atr = atrValue.ToDecimal();
				// Maintain running average for ATR
				_atrSum += atr;
				_atrCount++;
				decimal avgAtr = _atrCount > 0 ? _atrSum / _atrCount : atr;
				decimal volatilityRatio = Math.Min(Math.Max(atr / (candle.ClosePrice * 0.01m), 0), 1);

				// Higher volatility = shorter period and wider bands
				int newPeriod = MaxBollingerPeriod - (int)Math.Round(volatilityRatio * (MaxBollingerPeriod - MinBollingerPeriod));
				decimal newDeviation = MinBollingerDeviation + volatilityRatio * (MaxBollingerDeviation - MinBollingerDeviation);

				// Ensure parameters stay within bounds
				newPeriod = Math.Max(MinBollingerPeriod, Math.Min(MaxBollingerPeriod, newPeriod));
				newDeviation = Math.Max(MinBollingerDeviation, Math.Min(MaxBollingerDeviation, newDeviation));

				// Update Bollinger parameters if changed
				if (newPeriod != _currentBollingerPeriod || newDeviation != _currentBollingerDeviation)
				{
					_currentBollingerPeriod = newPeriod;
					_currentBollingerDeviation = newDeviation;

					_bollinger.Length = _currentBollingerPeriod;
					_bollinger.Width = _currentBollingerDeviation;

					LogInfo($"Adjusted Bollinger parameters: Period={_currentBollingerPeriod}, Deviation={_currentBollingerDeviation:F2} based on ATR={atr:F6}");
				}
			}

			// --- Bollinger logic (was ProcessBollinger) ---
			if (!IsFormedAndOnlineAndAllowTrading())
			return;

			if (bollingerValue.IsFinal && _atr.IsFormed)
			{
				decimal atrVal = atrValue.ToDecimal(); // use current ATR value
				// use running average
				bool isHighVolatility = atrVal > (_atrCount > 0 ? _atrSum / _atrCount : atrVal);

				var bollingerTyped = (BollingerBandsValue)bollingerValue;

				if (bollingerTyped.UpBand is not decimal upperBand ||
				bollingerTyped.LowBand is not decimal lowerBand ||
				bollingerTyped.MovingAverage is not decimal middleBand)
				{
					return; // Not enough data to calculate bands
				}

				if (isHighVolatility)
				{
					// Breakout above upper band - Sell signal
					if (candle.ClosePrice > upperBand && Position >= 0)
					{
						LogInfo($"Sell signal: Price ({candle.ClosePrice}) broke above upper Bollinger Band ({upperBand}) in high volatility");
						SellMarket(Volume + Math.Abs(Position));
					}
					// Breakout below lower band - Buy signal
					else if (candle.ClosePrice < lowerBand && Position <= 0)
					{
						LogInfo($"Buy signal: Price ({candle.ClosePrice}) broke below lower Bollinger Band ({lowerBand}) in high volatility");
						BuyMarket(Volume + Math.Abs(Position));
					}
				}

				// Exit logic based on middle band reversion
				if ((Position > 0 && candle.ClosePrice > middleBand) || 
				(Position < 0 && candle.ClosePrice < middleBand))
				{
					LogInfo($"Exit signal: Price ({candle.ClosePrice}) reverted to middle band ({middleBand})");
					ClosePosition();
				}
			}
		}
	}
}

File: API/0338_Adaptive_Bollinger_Breakout/PY/adaptive_bollinger_breakout_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")

from System import TimeSpan, Math
from StockSharp.Messages import DataType, Unit, UnitTypes, CandleStates
from StockSharp.Algo.Indicators import BollingerBands, AverageTrueRange
from StockSharp.Algo.Strategies import Strategy
from datatype_extensions import *
from indicator_extensions import *

class adaptive_bollinger_breakout_strategy(Strategy):
    """Strategy that trades based on breakouts of Bollinger Bands with adaptively adjusted parameters
    based on market volatility."""

    def __init__(self):
        super(adaptive_bollinger_breakout_strategy, self).__init__()

        # Strategy parameter: Minimum Bollinger period.
        self._min_bollinger_period = self.Param("MinBollingerPeriod", 10) \
            .SetGreaterThanZero() \
            .SetDisplay("Min Bollinger Period", "Minimum period for adaptive Bollinger Bands", "Indicator Settings")

        # Strategy parameter: Maximum Bollinger period.
        self._max_bollinger_period = self.Param("MaxBollingerPeriod", 30) \
            .SetGreaterThanZero() \
            .SetDisplay("Max Bollinger Period", "Maximum period for adaptive Bollinger Bands", "Indicator Settings")

        # Strategy parameter: Minimum Bollinger deviation.
        self._min_bollinger_deviation = self.Param("MinBollingerDeviation", 1.5) \
            .SetGreaterThanZero() \
            .SetDisplay("Min Bollinger Deviation", "Minimum standard deviation multiplier", "Indicator Settings")

        # Strategy parameter: Maximum Bollinger deviation.
        self._max_bollinger_deviation = self.Param("MaxBollingerDeviation", 2.5) \
            .SetGreaterThanZero() \
            .SetDisplay("Max Bollinger Deviation", "Maximum standard deviation multiplier", "Indicator Settings")

        # Strategy parameter: ATR period for volatility calculation.
        self._atr_period = self.Param("AtrPeriod", 14) \
            .SetGreaterThanZero() \
            .SetDisplay("ATR Period", "Period for ATR volatility calculation", "Indicator Settings")

        # Strategy parameter: Candle type.
        self._candle_type = self.Param("CandleType", tf(5)) \
            .SetDisplay("Candle Type", "Type of candles to use", "General")

        # Internal fields
        self._current_bollinger_period = 0
        self._current_bollinger_deviation = 0.0
        self._bollinger = None
        self._atr = None
        self._atr_sum = 0.0
        self._atr_count = 0

    @property
    def min_bollinger_period(self):
        return self._min_bollinger_period.Value

    @min_bollinger_period.setter
    def min_bollinger_period(self, value):
        self._min_bollinger_period.Value = value

    @property
    def max_bollinger_period(self):
        return self._max_bollinger_period.Value

    @max_bollinger_period.setter
    def max_bollinger_period(self, value):
        self._max_bollinger_period.Value = value

    @property
    def min_bollinger_deviation(self):
        return self._min_bollinger_deviation.Value

    @min_bollinger_deviation.setter
    def min_bollinger_deviation(self, value):
        self._min_bollinger_deviation.Value = value

    @property
    def max_bollinger_deviation(self):
        return self._max_bollinger_deviation.Value

    @max_bollinger_deviation.setter
    def max_bollinger_deviation(self, value):
        self._max_bollinger_deviation.Value = value

    @property
    def atr_period(self):
        return self._atr_period.Value

    @atr_period.setter
    def atr_period(self, value):
        self._atr_period.Value = value

    @property
    def candle_type(self):
        return self._candle_type.Value

    @candle_type.setter
    def candle_type(self, value):
        self._candle_type.Value = value

    def GetWorkingSecurities(self):
        return [(self.Security, self.candle_type)]

    def OnReseted(self):
        super(adaptive_bollinger_breakout_strategy, self).OnReseted()
        # Initialize adaptive parameters
        self._current_bollinger_period = self.max_bollinger_period  # Start with maximum period
        self._current_bollinger_deviation = self.min_bollinger_deviation  # Start with minimum deviation
        self._atr = None
        self._bollinger = None
        self._atr_sum = 0.0
        self._atr_count = 0

    def OnStarted(self, time):
        super(adaptive_bollinger_breakout_strategy, self).OnStarted(time)

        # Create ATR indicator for volatility measurement
        self._atr = AverageTrueRange()
        self._atr.Length = self.atr_period

        # Create Bollinger Bands indicator with initial parameters
        self._bollinger = BollingerBands()
        self._bollinger.Length = self._current_bollinger_period
        self._bollinger.Width = self._current_bollinger_deviation

        # Create subscription for candles
        subscription = self.SubscribeCandles(self.candle_type)

        # Bind indicators to subscription and start
        subscription.BindEx(self._atr, self._bollinger, self.ProcessIndicators).Start()

        # Add chart visualization
        area = self.CreateChartArea()
        if area is not None:
            self.DrawCandles(area, subscription)
            self.DrawIndicator(area, self._bollinger)
            self.DrawOwnTrades(area)

        # Start position protection with ATR-based stop-loss
        self.StartProtection(
            takeProfit=Unit(0),
            stopLoss=Unit(2, UnitTypes.Absolute)
        )
    def ProcessIndicators(self, candle, atr_value, bollinger_value):
        # Skip unfinished candles
        if candle.State != CandleStates.Finished:
            return

        # --- ATR logic (was ProcessAtr) ---
        if atr_value.IsFinal:
            atr = float(atr_value)
            # Maintain running average for ATR
            self._atr_sum += atr
            self._atr_count += 1
            avg_atr = self._atr_sum / self._atr_count if self._atr_count > 0 else atr
            volatility_ratio = float(max(min(atr / (candle.ClosePrice * 0.01), 1), 0))

            # Higher volatility = shorter period and wider bands
            new_period = self.max_bollinger_period - int(round(volatility_ratio * (self.max_bollinger_period - self.min_bollinger_period)))
            new_deviation = self.min_bollinger_deviation + volatility_ratio * (self.max_bollinger_deviation - self.min_bollinger_deviation)

            # Ensure parameters stay within bounds
            new_period = max(self.min_bollinger_period, min(self.max_bollinger_period, new_period))
            new_deviation = max(self.min_bollinger_deviation, min(self.max_bollinger_deviation, new_deviation))

            # Update Bollinger parameters if changed
            if new_period != self._current_bollinger_period or new_deviation != self._current_bollinger_deviation:
                self._current_bollinger_period = new_period
                self._current_bollinger_deviation = new_deviation

                self._bollinger.Length = self._current_bollinger_period
                self._bollinger.Width = self._current_bollinger_deviation

                self.LogInfo("Adjusted Bollinger parameters: Period={0}, Deviation={1:.2f} based on ATR={2:.6f}".format(
                    self._current_bollinger_period, self._current_bollinger_deviation, atr))

        # --- Bollinger logic (was ProcessBollinger) ---
        if not self.IsFormedAndOnlineAndAllowTrading():
            return

        if bollinger_value.IsFinal and self._atr.IsFormed:
            atr_val = float(atr_value)  # use current ATR value
            # use running average
            is_high_volatility = atr_val > (self._atr_sum / self._atr_count if self._atr_count > 0 else atr_val)


            if (
                bollinger_value.UpBand is None
                or bollinger_value.LowBand is None
                or bollinger_value.MovingAverage is None
            ):
                return  # Not enough data to calculate bands

            upper_band = float(bollinger_value.UpBand)
            lower_band = float(bollinger_value.LowBand)
            middle_band = float(bollinger_value.MovingAverage)

            if is_high_volatility:
                # Breakout above upper band - Sell signal
                if candle.ClosePrice > upper_band and self.Position >= 0:
                    self.LogInfo("Sell signal: Price ({0}) broke above upper Bollinger Band ({1}) in high volatility".format(
                        candle.ClosePrice, upper_band))
                    self.SellMarket(self.Volume + Math.Abs(self.Position))
                # Breakout below lower band - Buy signal
                elif candle.ClosePrice < lower_band and self.Position <= 0:
                    self.LogInfo("Buy signal: Price ({0}) broke below lower Bollinger Band ({1}) in high volatility".format(
                        candle.ClosePrice, lower_band))
                    self.BuyMarket(self.Volume + Math.Abs(self.Position))

            # Exit logic based on middle band reversion
            if (self.Position > 0 and candle.ClosePrice > middle_band) or \
               (self.Position < 0 and candle.ClosePrice < middle_band):
                self.LogInfo("Exit signal: Price ({0}) reverted to middle band ({1})".format(
                    candle.ClosePrice, middle_band))
                self.ClosePosition()

    def CreateClone(self):
        """!! REQUIRED!! Creates a new instance of the strategy."""
        return adaptive_bollinger_breakout_strategy()